src/Psl/IO/Reader.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\IO;
+
+use Psl;
+use Psl\Str;
+
+final class Reader implements ReadHandle
+{
+    private ReadHandle $handle;
+
+    private bool    $eof = false;
+    private string  $buffer = '';
+
+    public function __construct(ReadHandle $handle)
+    {
+        $this->handle = $handle;
+    }
+
+    public function getHandle(): ReadHandle
+    {
+        return $this->handle;
+    }
+
+    public function read(?int $max_bytes = null): string
+    {
+        Psl\invariant($max_bytes === null || $max_bytes >= 0, '$max_bytes must be null, or >= 0');
+
+        if ($this->eof) {
+            return '';
+        }
+
+        if ($this->buffer === '') {
+            $this->buffer = $this->handle->read();
+        }
+
+        if ($this->buffer === '') {
+            $this->eof = true;
+
+            return '';
+        }
+
+        if (null === $max_bytes || $max_bytes >= Str\Byte\length($this->buffer)) {
+            $buffer = $this->buffer;
+            $this->buffer = '';
+
+            return $buffer;
+        }
+
+        $buffer = $this->buffer;
+        $this->buffer = Str\Byte\slice($buffer, $max_bytes);
+
+        return Str\Byte\slice($buffer, 0, $max_bytes);
+    }
+
+    public function readUntil(string $suffix): ?string
+    {
+        $buf = $this->buffer;
+        $idx = Str\Byte\search($buf, $suffix);
+        $suffix_len = Str\Byte\length($suffix);
+        if ($idx !== null) {
+            $this->buffer = Str\Byte\slice($buf, $idx + $suffix_len);
+            return Str\Byte\slice($buf, 0, $idx);
+        }
+
+        do {
+            $chunk = $this->handle->read();
+            if ($chunk === '') {
+                $this->buffer = $buf;
+                return null;
+            }
+            $buf .= $chunk;
+        } while (!Str\Byte\contains($chunk, $suffix));
+
+        $idx = Str\Byte\search($buf, $suffix);
+        Psl\invariant($idx !== null, 'Should not have exited loop without suffix');
+        $this->buffer = Str\Byte\slice($buf, $idx + $suffix_len);
+        return Str\Byte\slice($buf, 0, $idx);
+    }
```

Comment:
i just spent 4 hours thinking this method was doing something wrong, but i was wrong.

*but* made it 17 lines long instead of 21, so i guess that's a plus.

Reply:
Looks good. Do you really need that 'should not happen' check at the end or is it to satisfy psalm's `null|string`?